### DISCUSSION TITLE
SparseTensor support in HeteroData

### DISCUSSION THREAD

**Discussion by Vid201** [pyg-team/pytorch_geometric#3206] [General]
Hi, I am trying to use HeteroData object on MAG240M from OGB-LSC and I am having some troubles (not sure what I am doing wrong). The first question would be if SparseTensor for edge_index (or adj_t maybe) in HeteroData is supported? I did not found any example in tests or docs.

**Comment by rusty1s** [comment]
Yes, this is supported. Each occurrence of `edge_index` can simply be replaced by a `SparseTensor` variant `adj_t`. This way, you can create the heterogeneous graph in a similar fashion as we do in our OGB mag240m example. Keep in mind that loading node features into memory will consume a lot of RAM.

**Reply by Vid201** [reply]
Cool! Yeah, I am aware of that, should not be a problem.

The error I am getting when sampling is `ERROR: Unexpected segmentation fault encountered in worker.` (guess it is related to underlying C++). To be sure it is not related to node features, I also tried the same approach on OGB_MAG. So if I convert all edge_index tensors to SparseTensor, the error appears. This is how I convert the edge_index:

```
edge_type = ('paper', 'rev_writes', 'author')
edge_index = data[edge_type].edge_index
row, col = edge_index
N = data[edge_type[0]].x.shape[0]
perm = (N * row).add_(col).numpy().argsort()
row = row[perm]
col = col[perm]
adj_t = SparseTensor(row=row, col=col, sparse_sizes=(data[edge_type[0]].x.shape[0], data[edge_type[2]].x.shape[0]), is_sorted=True)
del data[edge_type].edge_index
data[edge_type].adj_t = adj_t
```

**Reply by Vid201** [reply]
Hmm, the error seems to appear only when converting rev (reverse) edge_index.

**Reply by rusty1s** [reply]
I think there are a few issues with your code:
1. `N` needs to be defined based on `data[edge_type[1]]`
2. You need to transpose the `adj_t` matrix: `adj_t = adj.t()`

Note that the `ToSparseTensor()` transform should do this work for you already out-of-the-box.

**Reply by Vid201** [reply]
Thanks! ToSparseTensor is what I was looking for.

Does it make sense that performance is now affected:
1. no conversion and pass `batch.edge_index_dict` to model -> 5 it/s
2. convert to `SparseTensor` and pass `batch.edge_index_dict` (torch tensor) to model -> 2.5 it/s
3. convert to `SparseTensor` and pass `batch.adj_t_dict` (SparseTensor) to model -> 1.75 it/s

**Reply by rusty1s** [reply]
This shouldn't be the case :) To which example are you referring to?

**Reply by Vid201** [reply]
https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/to_hetero_mag.py - using NeighborLoader.

**Reply by rusty1s** [reply]
Let me check and come back to you :)

**Reply by Vid201** [reply]
Sure, thank you very much :)

**Comment by rusty1s** [comment]
For me, using `ToSparseTensor()` with `batch.adj_t_dict` is equally fast. I think for `NeighborLoader` there shouldn't be a drastic difference in runtime, as you cannot really exploit the memory-efficiency of `SparseTensor` in local subgraphs where `num_nodes ~ num_edges`.

It is a bit weird that using `SparseTensor` is so much slower for you. Are you running on CPU (since I'm at 20-25 it/s)? Note that `SparseTensor` and passing edges via `batch.edge_index_dict` is not recommended, as it will convert the `SparseTensor` representation back to the `edge_index` one.

**Edit:** On CPU, using `SparseTensor` is twice as fast for me. Mmh.

**Reply by Vid201** [reply]
Thanks for testing. I tried on the different machine with less overall CPU usage (some other things are also running on the first machine) and now the performance is the same in both cases with much higher it/s. I guess copying the `SparseTensor` from CPU to GPU is more CPU intensive? Don't know.

In all cases, the models were running on GPU.

Thanks again for the help!
